src/lib.rs

```diff
@@ -27,38 +34,100 @@
 //! assert_eq!(weak.count(), 0);
 //! ```
 
+use notify::NotifySender;
 use std::fmt::{self, Display, Formatter};
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 
+mod notify;
+
+pub use notify::{NotifyError, NotifyHandle, NotifyTimeoutError};
+
 /// Essentially an AtomicUsize that is clonable and whose count is based
-/// on the number of copies. The count is automatically updated on Drop.
+/// on the number of copies (and their size). The count is automatically updated on Drop.
+///
+/// If you want a weak reference to the counter that doesn't affect the count, see:
+/// [`WeakCounter`].
 #[derive(Debug)]
 pub struct Counter {
     counter: Arc<AtomicUsize>,
+    notify: Vec<NotifySender>,
     size: usize,
 }
 
-/// A 'weak' Counter that does not affect the count.
+/// A 'weak' [`Counter`] that does not affect the count.
 #[derive(Clone, Debug)]
-pub struct WeakCounter(Arc<AtomicUsize>);
+pub struct WeakCounter {
+    counter: Arc<AtomicUsize>,
+    notify: Vec<NotifySender>,
+}
 
-impl Counter {
-    pub fn new() -> Counter {
-        Counter::new_with_size(1)
+/// A builder for the [`Counter`].
+pub struct CounterBuilder {
+    counter: Arc<AtomicUsize>,
+    size: usize,
+    notify: Vec<NotifySender>,
+}
+
+impl CounterBuilder {
+    /// Change the specified size of the new [`Counter`]. This counter will add
+    /// `size` to the count, and will remove `size` from the count
+    /// when dropped.
+    pub fn size(mut self, v: usize) -> Self {
```

Comment:
Since it's a builder pattern I prefer the assignment to match 1:1 with the field, so the usage would look like:
```
Counter::builder().size(5).build()
```